### PR TITLE
[WIP] Integrate SkeletonObject into GD5

### PR DESCRIPTION
**Pull request by Wend1go**
Integrate @Lizard-13 's SkeletonObject extension into GD5.
In the first step this will have to do without a graphical editor and only load an already existing animation.

ToDo:
- [x] Expose the existing Object/Conditions/Actions/Expressions into the GD5 IDE
- [x] Expose object editor in GD5 IDE
- [x] Add a way to load already existing animations
- [ ] ~~Add an instance renderer for the scene editor~~
- [x] Load animations using the official jsonResource via drop down list
- [x] Documentation / Wiki article
- [x] Fix broken extension icon
- [ ] Testing
- [ ] Example
- [x] Add property to disable preloading of JSON (by default, json will be loaded at the beginning)

### CODE REVIEW

**Comment by Wend1go** [comment]
Everything seems to show up in the IDE now.
Will work on the load skeletal animation action next.

**Comment by blurymind** [comment]
You are an absolute legend! Thank you for working on this 🥇 😺 👍

**Comment by 4ian** [comment]
Looks like a good start! 

Lizard did actually a test game. I recompiled GD4 and exported it as a json file so you can try to open it in GD5.

[SkeletonTestAsJson.zip](https://github.com/4ian/GDevelop/files/3352483/SkeletonTestAsJson.zip)

Preview is working well on GD4!

image

A few things that come to my mind:
* There will probably be a need to create an instance renderer for the Skeleton objects. At first this can be a simple icon like it was done in GD4 - though a preview would be great but might be as complex as writing the runtime object.
* Hopefully with the additions you made, the extension should work good enough as it to run the example, let me know how it goes.
* This will need careful testing and polishing. This is a huge feature in terms of size so ideally we could make multiple example/test games for now (to be sure there is no bug and help to debug if there are some) and for the future (to help maintain this).
* If we get something working, we can mark this as experimental as the video object for a while.

**Comment by Wend1go** [comment]
@blurymind Well basically I just followed the steps @4ian wrote in the other PR. I haven't done anything fancy yet. ^^
@4ian Thanks for the test project. Since you already compiled GD4, would you mind posting screenshots of the object editor? Would be really interesting to see how it looks like.

**Comment by Wend1go** [comment]
Opening the existing project and starting a preview is working:
dragon

I'm currently trying to find my way around the extension.
Found the function that is used to load the skeleton file:
https://github.com/4ian/GDevelop/blob/01b7f81507b8fe880fefde2bbc65903cda13f2a8/Extensions/SkeletonObject/Gskeletonruntimeobject.js#L42
`objectData` seems to hold the details of the animation including the file path to the `skeletalDataFilename` and `textureDataFilename`.
The variables composing objectData seem to be declared in:
https://github.com/4ian/GDevelop/blob/master/Extensions/SkeletonObject/SkeletonObject.h#L45

So I'd need a way to load these. 
Unfortunately I don't really have a clue about how to access these parameters from within the
https://github.com/4ian/GDevelop/blob/master/Extensions/SkeletonObject/Extension.cpp

Also need to find a way to declare the paths to the source files of the animation.
Two possibilities come into my mind:
1) We add a very simple Object data editor that only holds input boxes / file chooser buttons to open the ressource/load the ressource from the file system
2) Load the Json and Image filed from the ressource manager inside a "load skeletal animation" action.

This is really tough, especially since it is half made in C++. Hope I didn't bite more than I can chew. :sweat_smile:

**Comment by blurymind** [comment]
I wish the extension was made in js like the one rpg maker mv is using

**Comment by 4ian** [comment]
> I wish the extension was made in js like the one rpg maker mv is using

Don't panic though, the C++ part is just a file describing the actions/conditions. It's exactly the same thing as in JS. If you can write extensions in JS, you can do it in C++ like this one :) I think it's quite readable if you look at the code. The only thing is that you have to recompile libGD.js - which is more complex but doable if you follow the README on Linux.
90% of this extension is JS :)

> The variables composing objectData seem to be declared in:

Yes they are data stored in the object.
See the object editor in GD5: 
image

By the way, are properties properly shown?

> Unfortunately I don't really have a clue about how to access these parameters from within the
https://github.com/4ian/GDevelop/blob/master/Extensions/SkeletonObject/Extension.cpp

You don't need to access to them in this file? It's just the file describing the extension? This file does not need to deal with the inside of the object.

> We add a very simple Object data editor that only holds input boxes / file chooser buttons to open the resources/load the resources from the file system

Yes that would be ideal. See again the "object editor" in GD4, it was only a few properties - which GD5 should be able to display, let me know if it's not the case.

You can see the object editor is already having something called "Texture data filename" and "Texture", and "Skeletal data filename".
These are the properties you've linked previously:

https://github.com/4ian/GDevelop/blob/0704d82348fdde8273e08951bdcc758c192e59ca/Extensions/SkeletonObject/SkeletonObject.h#L45-L48

and these properties, being properties of the object, are accessible in `objectData` in the runtime object (i.e: here https://github.com/4ian/GDevelop/blob/01b7f81507b8fe880fefde2bbc65903cda13f2a8/Extensions/SkeletonObject/Gskeletonruntimeobject.js#L42)

So that mean that the extension has been doing loading of JSONs, before they were even added as resources ;) 
Where is this done? Let's search for "textureDataFilename" in the files, we find this:

https://github.com/4ian/GDevelop/blob/0704d82348fdde8273e08951bdcc758c192e59ca/Extensions/SkeletonObject/Bskeletonruntimeobject-pixi-renderer.js#L21-L24

and getData is this:

https://github.com/4ian/GDevelop/blob/0704d82348fdde8273e08951bdcc758c192e59ca/Extensions/SkeletonObject/Bskeletonruntimeobject-pixi-renderer.js#L14-L19

So what Lizard-13 is simply to add fake "images" resources, that are JSONs, are got them from PIXI (as all "images" are loaded at the beginning of the game). This is a bit "hacky" but it works :) Going forward, if we want to do this properly, we'll use `jsonResource` as resources in the object I think. How do to this?
Here is an example in the Video object:

https://github.com/4ian/GDevelop/blob/0704d82348fdde8273e08951bdcc758c192e59ca/Extensions/Video/JsExtension.js#L74-L77

Here, we'll have to do it in the C++ file, but it's mostly similar. Here precisely:
https://github.com/4ian/GDevelop/blob/0704d82348fdde8273e08951bdcc758c192e59ca/Extensions/SkeletonObject/SkeletonObject.cpp#L52-L66

The current fields are just raw strings, while we want them to be selector of resources. I might have to rework slightly json resources to allow them to be specified as "load at the start".

So to summarize:
- See if the object editor is working or not already. If not, let's see why (by default, it should show the properties). Should be 100% JS.
- Let's change the properties to be selector for resources (we can actually do this later). A bit of C++ involved.
  - Surely we'll have to adapt a few things to load properly the jsons. Should be 100% JS.
- Test all the extensions actions and so on to see if they all work and are usable :)
- ???
- Profit! 😄

**Comment by 4ian** [comment]
We might be actually closer to something working than expected. The thing is that I'm not an advanced users of these animation software like DragonBones, so it's hard for me to develop and maintain such an extension. If you can test it, make examples, check for bugs, check that it is working for most usages, then it would greatly help :)

**Comment by blurymind** [comment]
I used to work as an animator and have some knowledge of dragonbones from
other engines. The official dragonbones git repo has many example files
that we can easily use for testing when implementing this. I believe I
linked to it on Trello.

Could help out with testing this. I noticed on the screenshot that some of
the body parts seem off a bit. Like they are positioned with a strange
offset.

A big win we can get out of this is the ability to reuse animation on
different assets, basically by swapping the image atlas during runtime. The
other one is in the ability to smoothly interpolate between animations and
even mix multiple animations or animations with user input driving rotation
(aim gun mechanic)

**Comment by Wend1go** [comment]
Thanks for the explanation. Unfortunately I was unable to open the Object editor dialog.
Will look into it at the weekend if I find some time.

**Comment by 4ian** [comment]
Well I forgot that you had to register an object editor :)
Do it here: https://github.com/4ian/GDevelop/blob/master/newIDE/app/src/ObjectEditor/ObjectsEditorService.js#L104

* Use `ObjectPropertiesEditor` as the `component` (which is what is done for JS extensions).
* For the function to create an object: `createNewObject: () => new gd.SkeletonObject(''),` - there is no gd.SkeletonObject in libGD.js so this won't work yet but keep reading.
* For the function to cast an object (used for the IDE to "transform" from a generic object into an object of this type): `castToObjectType: object => gd.castObject(object, gd.SkeletonObject),` (the gd.asXXX methods are [declared here](https://github.com/4ian/GDevelop/blob/master/GDevelop.js/Bindings/postjs.js#L92) in libGD.js, which is a design mistake from me because it just make things harder to understand.)

For gd.SkeletonObject, you'll need to expose the object in the Bindings (to let the IDE, in JS, create a C++ object: 

1. In Bindings.idl, do like the TextEntryObject, but with the name of the C++ object ():
https://github.com/4ian/GDevelop/blob/0ad06581ec6328e2ceae53afd347b46ec19f43d5/GDevelop.js/Bindings/Bindings.idl#L2075-L2078
2. To compile properly, the bindings need to know about the C++ class header file (`Extensions/SkeletonObject/SkeletonObject.h`). Add it here:
https://github.com/4ian/GDevelop/blob/0ad06581ec6328e2ceae53afd347b46ec19f43d5/GDevelop.js/Bindings/Wrapper.cpp#L71
3. Recompile libGD.js (`npm run build`)
4. Done! You know have a `gd.SkeletonObject`! It's only going to be useful for writing this tiny code to expose an object editor (this is necessary because the object editor has to be able to create a new object and clone existing ones, so it needs to know the class, which is what you're exposing with createNewObject and castToObjectType

After this, the object editor should work and display the 4 properties that have been declared in C++. They should be displayed using a string field, which we can improve later.

**Comment by Wend1go** [comment]
It is working
grafik

**Comment by Wend1go** [comment]
When I create an new project and try to recreate the dragon animation it just shows nothing when I hit run. Not even errors in the console.
Looks like the "animation" property is null. Setting an animation via action doesn't help either.

**Comment by blurymind** [comment]
perhaps there is something in the project.json file that is is in the gd4 format and gd5 is either failing to write or read it in the required way?

**Comment by 4ian** [comment]
> Looks like the "animation" property is null. Setting an animation via action doesn't help either

Could you post the json file of the game that is not working?

**Comment by Wend1go** [comment]
Here it is:
[SkeletonCreationEmpty.zip](https://github.com/4ian/GDevelop/files/3378504/SkeletonCreationEmpty.zip)

**Comment by 4ian** [comment]
Sorry did not have the time to check, will try to look next week!

**Comment by Wend1go** [comment]
No problem, looks like my PC has passed away so It'll take some time to get a new one and install everything. :frowning:

**Comment by 4ian** [comment]
Ah, sad to hear, hope you can get something working and have not lost too many things!

**Comment by Wend1go** [comment]
While I'm able to install and start the master branch, the skeleton branch still refuses to start after reinstalling my pc.
GDevelop.js compiles fine installing newIDE/app too but when I run `npm start`, I get the following error. (deleting node_modules doesn't help either).
Do you have any ideas what could cause this? `webpack`isn't even part of the package.json.
```
me@mypc:~/Dokumente/GD_MyFork/GDevelop/newIDE/app$ npm start

> gdevelop-ide@0.1.0 start /home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app
> npm run import-resources && react-scripts start


> gdevelop-ide@0.1.0 import-resources /home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app
> cd scripts && node import-libGD.js && node import-res-folder.js && node import-GDJS-Runtime.js && node import-monaco-editor.js && node import-zipped-editor.js piskel 5.0.0-beta56 a2c36775109a1c1181d0b9ab9b7c903365c4ecb340118f1237b66236f23e20dd && node import-zipped-editor.js jfxr 5.0.0-beta55 8ac12b557c2ddba958c6f0d3e0c5df8cf3369a65262dcb90cf5c8a7a7d20bdf6

✅ Copied libGD.js from Binaries/Output/libGD.js/Release to public and node_modules folder
✅ Images from res, CppPlatform and JsPlatform properly copied in public folder
ℹ️ Copying GDJS and extensions runtime files (*.js) to '/home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app/resources/GDJS/Runtime'...
✅ Copied GDJS and extensions runtime files (*.js) to '/home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app/resources/GDJS/Runtime'.
ℹ️ Copying GDJS and extensions runtime files (*.js) to '/home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app/node_modules/GDJS-for-web-app-only/Runtime'...
✅ Copied GDJS and extensions runtime files (*.js) to '/home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app/node_modules/GDJS-for-web-app-only/Runtime'.
✅ Sources of Monaco Editor properly copied in public folder
✅ piskel-editor already existing in public/external/piskel folder and up-to-date - skipping download
✅ jfxr-editor already existing in public/external/jfxr folder and up-to-date - skipping download

There might be a problem with the project dependency tree.
It is likely not a bug in Create React App, but something you need to fix locally.

The react-scripts package provided by Create React App requires a dependency:

  "webpack": "4.28.3"

Don't try to install it manually: your package manager does it automatically.
However, a different version of webpack was detected higher up in the tree:

  /home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app/node_modules/webpack (version: 4.39.1) 

Manually installing incompatible versions is known to cause hard-to-debug issues.

If you would prefer to ignore this check, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
That will permanently disable this message but you might encounter other issues.

To fix the dependency tree, try following the steps below in the exact order:

  1. Delete package-lock.json (not package.json!) and/or yarn.lock in your project folder.
  2. Delete node_modules in your project folder.
  3. Remove "webpack" from dependencies and/or devDependencies in the package.json file in your project folder.
  4. Run npm install or yarn, depending on the package manager you use.

In most cases, this should be enough to fix the problem.
If this has not helped, there are a few other things you can try:

  5. If you used npm, install yarn (http://yarnpkg.com/) and repeat the above steps with it instead.
     This may help because npm has known issues with package hoisting which may get resolved in future versions.

  6. Check if /home/me/Dokumente/GD_MyFork/GDevelop/newIDE/app/node_modules/webpack is outside your project directory.
     For example, you might have accidentally installed something in your home folder.

  7. Try running npm ls webpack in your project folder.
     This will tell you which other package (apart from the expected react-scripts) installed webpack.

If nothing else helps, add SKIP_PREFLIGHT_CHECK=true to an .env file in your project.
That would permanently disable this preflight check in case you want to proceed anyway.

P.S. We know this message is long but please read the steps above :-) We hope you find them helpful!

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! gdevelop-ide@0.1.0 start: `npm run import-resources && react-scripts start`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the gdevelop-ide@0.1.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/me/.npm/_logs/2019-08-11T07_03_18_001Z-debug.log
```

**Comment by Wend1go** [comment]
I compared the two project files and found the following differences:
1) `properties/extensions` is missing the SkeletonObject:
```
            {
                "name": "SkeletonObject"
            },
```
2) `resources/resources` is missing the following:
```
            {
                "alwaysLoaded": false,
                "file": "skeleton.json",
                "kind": "image",
                "metadata": "",
                "name": "skeleton.json",
                "smoothed": true,
                "userAdded": false
            },
            {
                "alwaysLoaded": false,
                "file": "skeleton_texture.json",
                "kind": "image",
                "metadata": "",
                "name": "skeleton_texture.json",
                "smoothed": true,
                "userAdded": false
            },
            {
                "alwaysLoaded": false,
                "file": "skeleton_texture.png",
                "kind": "image",
                "metadata": "",
                "name": "skeleton_texture.png",
                "smoothed": true,
                "userAdded": false
            }
```

**Comment by 4ian** [comment]
Is master working? 
I.e: if you go on master, delete node_modules, run npm install and npm start, does it work?

The issue is with dependencies installation, not the project files or anything in the source code.

If yes: might be something with your branch package.json/package-lock.json
If no: try install and using yarn instead of npm (remove node_modules, yarn install, yarn start)

**Comment by Wend1go** [comment]
It is all working, already deleted the first post. Merging in 4ian/master did the trick.
I'm already a step further, found the difference in the project files (see above)

**Comment by 4ian** [comment]
Sorry for the late reply, so seems that the issue is that the resources are not added. This is indeed "normal" because the properties exposed by the SkeletonObject (in SkeletonObject.cpp) are string properties, so they are not triggering the import of any resource (and they are displayed as a text field, not as a resource selector).

We would need "Skeletal data filename" and "Texture data filename" to be a understood by the editor as field asking for a "json resource" (like used by @blurymind extension for dialogue tree) and "Texture" to be an "image resource".
How do to it? There is an example in the video object done by @Bouh:

https://github.com/4ian/GDevelop/blob/bc80d1c98f22bb8aae559b1fbcb25452d9a6b9d8/Extensions/Video/JsExtension.js#L72-L78

By setting a property with a type "resource", with an extra info containing the name of the resource type ("image", "audio", "font" or "json"), the editor will automatically* create a field that will import the resource in the game.

So can you try to modify SkeletonObject.cpp, to change: 

```
  properties[_("Skeletal data filename")].SetValue(skeletalDataFilename);
  properties[_("Main armature name")].SetValue(rootArmatureName);
  properties[_("Texture data filename")].SetValue(textureDataFilename);
  properties[_("Texture")].SetValue(textureName);
```

to:

```
  properties[_("Skeletal data filename")].SetValue(skeletalDataFilename)
    .SetType("resource").AddExtraInfo("json");
  properties[_("Main armature name")].SetValue(rootArmatureName);
  properties[_("Texture data filename")].SetValue(textureDataFilename)
    .SetType("resource").AddExtraInfo("json");
  properties[_("Texture")].SetValue(textureName)
    .SetType("resource").AddExtraInfo("image");
```

?

Then recompile libGD.js (npm run build), refresh the editor and see if when you add/edit a skeleton object, the field are properly shown as selectors. If yes, you should be able to recreate a new project from scratch and have it working.

Let me know if it works :) 

* precisely, if you want to know how it's done, the properties are mapped [here](https://github.com/4ian/GDevelop/blob/bc80d1c98f22bb8aae559b1fbcb25452d9a6b9d8/newIDE/app/src/PropertiesEditor/PropertiesMapToSchema.js#L109-L124) to a "schema", and the "schema" is [used in PropertiesEditor](https://github.com/4ian/GDevelop/blob/bc80d1c98f22bb8aae559b1fbcb25452d9a6b9d8/newIDE/app/src/PropertiesEditor/index.js#L314) to render a field that is a `ResourceSelector`.

**Comment by 4ian** [comment]
I've just written some documentation about properties here in case it's useful in the future: https://github.com/4ian/GDevelop/blob/master/newIDE/docs/Properties-schema-and-PropertiesEditor-explanations.md :)

**Comment by blurymind** [comment]
Thank you @4ian :) perhaps it will be good to add what is not supported atm

Btw today an interesting project similar to gd was open sourced
https://github.com/ct-js/ct-js
The reason I mention it is because:
- uses pixijs for rendering
- has a dragonbones extension built in 
- It directly integrates the official dragonbones js extension - straight from the dragonbones repository and adds a layer to the engine to communicate with it. It's purely javascript. Just mentioning in case this one is too problematic to maintain :) a few of us are as good with c++ as  js

**Comment by 4ian** [comment]
Yes, if there are more use cases I'll take a look at adding them :)

ct.js looks interesting. I'm taking a look - thanks for letting me know :)
How did you find about it?

I'm pretty sure this dragonbone integration in GDevelop can be as good as
something in another engine.

This being said, if someone want to integrate dragonbone library in
GDevelop, you're very welcome - GDevelop has everything ready as we can see
in this Skeleton extension - the only difference is that Lizard-13 went
into making a skeleton implementation by himself while we could instead
leverage a library like dragonbones.
Advantage: it's already battle tested. Disadvantage: it needs a layer to
integrate itself into GD game engine (collisions, create a runtime object
on top of it).

**Comment by 4ian** [comment]
Here, we're literally **a few lines of code** away of having something
working!
If my previous explanations are correct, it's a matter of fixing the field
to have resource selector (which is 3 lines of C++ that are exactly the
same as what you would write in JS), recompile GDevelop.js and we're done -
we have a working skeleton object that can read dragonbone files. 😃

(If you're on Linux, I encourage you to try to compile GDevelop.js. It's
probably way easier than what you think! :) f you're on Windows I've
unfortunately not tried it on Windows recently - but happy to help fixing
this.)

@Wen1go, let me know if you have time, otherwise happy to take a look at
this.

**Comment by 4ian** [comment]
I went ahead and made the changes I was speaking about :)
This is still not working when creating a new skeleton from scratch because the extension is using a way to load JSON that was a "hack" on top of the image resources (it was asking you to add an image that was a json to be able to load it). We should use the new json manager from GDevelop but it's asynchronous - I may have to add the possibility for json to be loaded at the same time as the game images/audio/fonts.
And also add `console.error` when there is an error during loading - otherwise it's hard to know what was wrong :)

Btw, good news, all of this is in JS ;)

**Comment by 4ian** [comment]
It's working :) I had to add preloading of JSON resources (but I think that in the case of @blurymind's Dialogue Tree extension, it's better to keep the asynchronous loading that we did).
Then I updated the extension to use `gdjs.JsonManager` and here we go:

image

I'll merge this extension with an "{experimental)" flag :)

**Comment by 4ian** [comment]
Well, better add at least an icon for the object on the scene editor before merging this :)

**Comment by Wend1go** [comment]
Haha, you were faster than me. ^^
Should we already merge this (when the icon is ready) or implement the instance renderer first?
Would be pretty hard to place the DB object precisely in the scene by only using a small icon.

**Comment by 4ian** [comment]
I added an icon for now - very simple but does the job.

> Should we already merge this (when the icon is ready) or implement the instance renderer first?

I'm not sure. The instance renderer is tricky because the object is quite complex, so we would almost need a copy of the whole source code of the object to render an instance :/ Would maybe have been simpler with a lib like the one from dragonbones though I'm not really sure.

Do you think there is a way to show a simplified version of the skeleton?

**Comment by 4ian** [comment]
The DragonBonesJS library seems relatively easy to use: https://github.com/DragonBones/DragonBonesJS/blob/master/Pixi/Demos/src/HelloDragonBones.ts

Would be interesting to compare what their implementation does compared to the current manual implementation of the Skeleton object. Advantage of using the DragonBonesJS would be to use their "official" library, which I guess is 100% compatible with all their features and we can offload all the complexity of the implementation into this lib. On the other hand, we would have to handle things like collisions - and if we want to switch to another skeletal animation library, we have to create a new object type (but this is ok I think).

This could simplify the rendering of instances too, because we could load the jsons and call the library like in the example I linked.

**Comment by blurymind** [comment]
@4ian thank you for working on this feature. The main advantage of using the 🐉 bones official one is the support that we will be getting as well as new features and bugfixes from the community. Lizard13 did a fantastic job with his implementation, but his time is a limited resource and his implementation can only be us d in gdevelop, making it less supported by the wider open source community. I think other engines who decided to pick up the official library made a smart move as it freed their own developers from the burden of maintaining the entire functionality.

In any case, I am not saying this to discourage this. I am very much excited to see this in gdevelop and think it will definitely attract a lot of new users to the engine :) 

For CT.js I discovered it on itch.io. The author regularly posted updates there, which pushed his entry in the tools section up and made it noticeable. I really like how simple JavaScript scripting is in it. He split the code execution into tabs a bit like game maker. The code editor having it's own window and proper syntax coloring is nice. Unfortunately doesn't have alot of the awesome things GD does, but is still a cool new entry

**Comment by Wend1go** [comment]
Hard to say. I have never worked on an instance renderer before.
Looking at the `skeleton.json` file it looks like each animation has an `aabb` property. If we pick the one from the first animation to scale the icon in the scene to the correct bounding box dimensions it might be sufficient.
Or if it is too much hassle lets just publish it

**Comment by blurymind** [comment]
Any updates on this? I would be fine if we dont have a renderer in the editor in the start. It's more important that it works in the runtime well

**Comment by 4ian** [comment]
> One idea is If this one turns out to be difficult to maintain, we might give it a try later on to create one with the official dragonbones library and have it alongside this one - with this one called "Skeleton Object" and the other one being called "Dragonbones Object".

This is my main fear. I would not like to introduce an object that is going to be unmaintained or create lots of bug reports not addressed :/  (not saying that it's going to be the case, but bad things™️ will happen).

**Comment by blurymind** [comment]
> > One idea is If this one turns out to be difficult to maintain, we might give it a try later on to create one with the official dragonbones library and have it alongside this one - with this one called "Skeleton Object" and the other one being called "Dragonbones Object".
> 
> This is my main fear. I would not like to introduce an object that is going to be unmaintained or create lots of bug reports not addressed :/ (not saying that it's going to be the case, but bad things will happen).

Looking at the recent past, we have had fewer c++ contributors than javascript ones. It's simply harder to develop in c++

**Comment by 4ian** [comment]
It's not the C++ side that is making me afraid here - if that's the only thing we can convert the extension declaration to JS in a day - but the fact that this is an entirely home-made implementation of a skeleton object in 9 files :) That's quite a lot of code to maintain, and with specific knowledge to have about skeleton.

**Comment by blurymind** [comment]
Yes that as well.. Granted it works and is better than having nothing - it is a fantastic feature. Perhaps if we can some day replace it with a dragonbones library pure pixi implementation, we can phase the skeleton out.

Not sure how we will make the collision work, but perhaps as a starting point it would be great to simply expose its features to the event sheet and gdevelop itself. It's already being used by a number of other game engines btw

**Comment by blurymind** [comment]
One way to get more people interested in the skeleton object code is to have it work as a third party library different engines can adopt - on a separate repository. If more engines than gdevelop use it, it is more likely to receive maintenance updates from a wider community. I really don't wan't to see the fantastic work @Lizard-13 has done go unmaintained

**Comment by 4ian** [comment]
This would be a good idea, though the separation might involve refactoring of the part that are dealing with GDJS game engine (collision, etc).
Using a third party library would still need work to integrate these specific points (wrap the thing into an object, handle collisions, etc...) but hopefully the said third party library would be regularly maintained/improved (there is not only bugs to fix, but also potential performance issues or these kind of things that can happen) and we would only have a "thin layer" exposing it as an object.

**Comment by 4ian** [comment]
I'll merge this this week and add a warning in the editor about this object so that user know that they might have to migrate in the future.

**Comment by 4ian** [comment]
I need to add `disablePreload` property to JSON resource, as json used by Yarn might not want to be all loaded at the beginning of the game (though it will be a possibility when this is merged, in the case by default)

**Comment by 4ian** [comment]
Merged! Thanks for working on this :) This is flagged as experimental and I added banner to make clear that it's not actively maintained though..

**Comment by blurymind** [comment]
How does one use it? I dont see it listed with the objects to add

**Comment by blurymind** [comment]
This is what I get when I run the demo
error
a black screen
